### ISSUE TITLE
Investigate: switch to different DateTime library than moment (e.g., dayjs)

### ISSUE THREAD

**Issue by jeromesimeon** [accordproject/concerto#205]
moment.js is nice but big and with lots of features we do not use.

We may want to consider switching to a different library which is smaller, if we can. e.g., https://github.com/iamkun/dayjs

https://moment.github.io/luxon/

### Also

We should probably do this anyway, considering the recently posted status of development for moment.js:
https://momentjs.com/docs/#/-project-status/

**Comment by DianaLease** [comment]
Slides comparing Moment vs Dayjs vs Luxon: https://docs.google.com/presentation/d/1bGSR076MzrLslwMMdGPkEg4Hyf_PZDad5kEeV5g5r-E/edit?usp=sharing

Repl comparing usage & performance: https://repl.it/@DianaLease/Comparing-Moment-Alternatives#perf-tests.js

**Comment by irmerk** [comment]
Went into the call favoring Luxon by a lot, but I think the case was made for Dayjs and we should definitely switch to that. That is, given that we feel the plugins cover our use cases and vanilla JS isn't our pick.

**Comment by jeromesimeon** [comment]
> Went into the call favoring Luxon by a lot, but I think the case was made for Dayjs and we should definitely switch to that. That is, given that we feel the plugins cover our use cases and vanilla JS isn't our pick.

Thanks for the analysis. I think after today's presentation I agree.

I gave a quick look at https://date-fns.org It looks nice, but is bigger than day.js -- haven't gotten a chance to look at performances and features. One possible benefit is that it relies on JavaScript Date objects so we could, I believe, completely remove dependency to date/time libraries in Concerto, and use it only for evaluation in Ergo.

**Comment by DianaLease** [comment]
List of `moment` methods used in the `concerto-core` package of the release-1.0 branch:
- `Moment().utc()`
- `Moment.parseZone(...)`
- `Moment.isMoment(...)`
- `moment(..., moment.ISO_8601).isValid()`
- `new Moment()`

**Comment by jeromesimeon** [comment]
> List of `moment` methods used in the `concerto-core` package of the release-1.0 branch:
> 
> * `Moment().utc()`
> * `Moment.parseZone(...)`
> * `Moment.isMoment(...)`
> * `moment(..., moment.ISO_8601).isValid()`
> * `new Moment()`

A small test to explain each of those:
```
bash-3.2$ node test.js
new Moment()                                  : 2020-11-04T15:54:40-05:00
Moment.utc()                                  : 2020-11-04T20:54:40Z
const date = '2020-11-04T20:31:09.007-05:00';
Moment(date)                                  : 2020-11-04T20:31:09-05:00
Moment(date, moment.ISO_860).isValid()        : true
Moment('foobar', moment.ISO_8601).isValid()   : false
Moment.parseZone(date)                        : 2020-11-04T20:31:09-05:00
Moment.isMoment(0)                            : false
Moment.isMoment(date)                         : false
```

**Comment by DianaLease** [comment]
We also call `.format()` on Moment date objects.
e.g. `obj.format('YYYY-MM-DDTHH:mm:ss.SSS[Z]')` and `obj.format('YYYY-MM-DDTHH:mm:ss.SSSZ')` where `obj` is a Moment date object

**Comment by DianaLease** [comment]
Reminder that `date-fns` (the library that uses the functional programming approach) has been added to the repl, and I have just now also added it the slides. Same links:
- Slides comparing Moment vs alternatives: https://docs.google.com/presentation/d/1bGSR076MzrLslwMMdGPkEg4Hyf_PZDad5kEeV5g5r-E/edit?usp=sharing
- Repl comparing usage & performance: https://repl.it/@DianaLease/Comparing-Moment-Alternatives#perf-tests.js

**Comment by K-Kumar-01** [comment]
@DianaLease @jeromesimeon @irmerk 
>Repl comparing usage & performance: https://repl.it/@DianaLease/Comparing-Moment-Alternatives#perf-tests.js

Here `dayjs` is better than `moment` in some cases and vice versa.
Also `date-fns` is way better in performance than others.

However, the size of `date-fns` is bigger than both `moment` as well as `dayjs`.

As `dayjs` has a similar API to `moment` we can try that as the migration will be easier in this case.

If the issue is still open, I would like to have a go at it. 
I am thinking of using `dayjs` though I can try others if needed.

**Comment by jeromesimeon** [comment]
Resolved and implemented in #218.
